File: hydra-haskell/src/main/haskell/Hydra/Lib/Strings.hs
-- | Haskell implementations of hydra.lib.strings primitives

module Hydra.Lib.Strings where

import qualified Data.Char as C
import qualified Data.List as L
import qualified Data.List.Split as LS


cat :: [String] -> String
cat = L.concat

cat2 :: String -> String -> String
cat2 s1 s2 = s1 ++ s2

-- | Get the Unicode code point at the given index. Implements hydra.lib.strings.charAt.
-- TODO: partial function. See https://github.com/CategoricalData/hydra/issues/201
charAt :: Int -> String -> Int
charAt i s = C.ord (s !! i)

-- | Convert a list of Unicode code points to a string. Implements hydra.lib.strings.fromList.
fromList :: [Int] -> String
fromList = fmap C.chr

intercalate :: String -> [String] -> String
intercalate = L.intercalate

length :: String -> Int
length = L.length

lines :: String -> [String]
lines = L.lines

null :: String -> Bool
null = L.null

splitOn :: String -> String -> [String]
splitOn = LS.splitOn

-- | Convert a string to a list of Unicode code points. Implements hydra.lib.strings.toList.
toList :: String -> [Int]
toList = fmap C.ord

toLower :: String -> String
toLower = fmap C.toLower

toUpper :: String -> String
toUpper = fmap C.toUpper

unlines :: [String] -> String
unlines = L.unlines

File: packages/hydra-rust/src/hydra/lib/strings.rs
// hydra.lib.strings primitives

pub fn cat(ss: Vec<String>) -> String {
    ss.concat()
}

pub fn cat2(s1: String, s2: String) -> String {
    format!("{}{}", s1, s2)
}

pub fn from_list(cs: Vec<i32>) -> String {
    cs.into_iter().filter_map(|c| char::from_u32(c as u32)).collect()
}

pub fn intercalate(sep: String, ss: Vec<String>) -> String {
    ss.join(&sep)
}

pub fn is_empty(s: String) -> bool {
    s.is_empty()
}

pub fn length(s: String) -> i32 {
    s.len() as i32
}

pub fn lines(s: String) -> Vec<String> {
    s.lines().map(String::from).collect()
}

pub fn split_on(sep: String, s: String) -> Vec<String> {
    s.split(&sep).map(String::from).collect()
}

pub fn to_list(s: String) -> Vec<i32> {
    s.chars().map(|c| c as i32).collect()
}

pub fn to_lower(s: String) -> String {
    s.to_lowercase()
}

pub fn to_upper(s: String) -> String {
    s.to_uppercase()
}

pub fn unlines(ss: Vec<String>) -> String {
    ss.join("\n")
}
